Sources/Algorithms/LazySplitCollection.swift

```diff
@@ -0,0 +1,367 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// A collection that lazily splits a base collection into subsequences
+/// separated by elements that satisfy the given `whereSeparator` predicate.
+///
+/// - Note: This type is the result of
+///
+///     x.split(maxSplits:omittingEmptySubsequences:whereSeparator)
+///     x.split(separator:maxSplits:omittingEmptySubsequences)
+///
+///   where `x` conforms to `LazyCollectionProtocol`.
+public struct LazySplitCollection<Base: Collection> {
+  internal let base: Base
+  internal let isSeparator: (Base.Element) -> Bool
+  internal let maxSplits: Int
+  internal let omittingEmptySubsequences: Bool
+  internal var _startIndex: Index
+
+  internal init(
+    base: Base,
+    isSeparator: @escaping (Base.Element) -> Bool,
+    maxSplits: Int,
+    omittingEmptySubsequences: Bool
+  ) {
+    self.base = base
+    self.isSeparator = isSeparator
+    self.maxSplits = maxSplits
+    self.omittingEmptySubsequences = omittingEmptySubsequences
+    self._startIndex = Index(baseRange: base.startIndex..<base.startIndex)
+
+    if !base.isEmpty {
+      // Precompute the start index.
+      _startIndex = indexForSubsequence(atOrAfter: base.startIndex)
+    }
```

Comment:
Nit: That way we avoid create an index that would be discarded if the condition `!base.isEmpty` is `true`
```suggestion
    if !base.isEmpty {
      // Precompute the start index.
      self._startIndex = indexForSubsequence(atOrAfter: base.startIndex)
    } else {
      self._startIndex = Index(baseRange: base.startIndex..<base.startIndex)
    }
```

Reply:
> Nit: That way we avoid create an index that would be discarded if the condition `!base.isEmpty` is `true`

When I try this, I get `Variable 'self._startIndex' used before being initialized` on line 40. I actually tried a similar thing first and ran into the same error, which is why it ended up the way it's committed now. I have a vague notion the compiler sees some difference between assigning the result of an initializer, and copying the result of a method call? Seems like a thing I should know. I'll research it.

Reply:
Ah sorry my mistake, the compiler is correct in that case ... I just didn't notice that `indexForSubsequence` is a member function of `self` so calling `self._startIndex = self.indexForSubsequence(atOrAfter: base.startIndex)` before `self._startIndex` initialization is definitely an access to `self` before all properties being initialized... so I think we have to stay with this

Reply:
Ah, right. Thanks for that explanation. I think the specificity of the error message confused me: `self` is definitely used before being initialized there, but `self._startIndex` isn't used in that member function.

Reply:
Ah this is just diagnostics that could be improved on the compiler :) 
We have a bug tracking similar situations where diagnostics could be more specific https://bugs.swift.org/browse/SR-12421